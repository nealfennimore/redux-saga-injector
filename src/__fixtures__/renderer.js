import PropTypes from 'prop-types';
import React from 'react';
import ReactDOMServer from 'react-dom/server';

import { Provider } from 'react-redux';
import createStore from '../__fixtures__/store';
import { preloadQueue } from '../sagas/queue';

const Application = ( {
    store,
    children,
} )=> (
    <Provider store={store}>
        { children }
    </Provider>
);

Application.propTypes = {
    store: PropTypes.shape( {} ).isRequired,
    children: PropTypes.node.isRequired,
};


const renderMarkup = ( App, {store} )=> (
    ReactDOMServer.renderToString(
        <Application store={store}>
            <App store={store} />
        </Application>
    )
);

export default function render( App, store, req, res, options ) {
    return async function() {
        // Server saga listens for any injected sagas to finish
        const preload = store.runSaga( preloadQueue, options );

        // Start initial render to start sagas
        // This is a throw away render
        renderMarkup( App, { store } );

        // Proceed after preload saga finishes
        await preload.done;

        // Get markup with updated store
        const html = renderMarkup( App, { store } );

        res.send( `
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id="app">${html}</div>
                </body>
            </html>
        ` );
    };
}
