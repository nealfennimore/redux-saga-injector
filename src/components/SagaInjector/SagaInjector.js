import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { cancelSagas, runSagas } from 'src/actions/sagas';
import { curry } from 'src/utils/fp';
import { genUID } from 'src/utils/uid';

/**
 * Inject sagas to run with component lifecycle
 *
 * @param {Object} options Should contain sagas as an array
 * @param {any} WrappedComponent Component to wrap
 * @returns SagaInjector
 */
function SagaInjectorHOC( options, WrappedComponent ) {
    const { sagas, STORE_KEY } = Object.assign( { sagas: [], STORE_KEY: 'store' }, options );

    const componentName = WrappedComponent.displayName ||
	WrappedComponent.name ||
    'Component';

    class SagaInjector extends Component {
        static displayName = `SagaInjector(${ componentName })`
        static propTypes = {
            [ STORE_KEY ]: PropTypes.object // eslint-disable-line
        }
        static contextTypes = {
            [ STORE_KEY ]: PropTypes.object
        }

        state = {
            injectedSagas: false,
        }

        uid = genUID()
        sagas = sagas

        componentWillMount() {
            this.injectSagas();
        }

        componentWillUnmount() {
            this.store.dispatch( cancelSagas( this.uid ) );
        }

        get store() {
            return this.props[STORE_KEY] || this.context[STORE_KEY];
        }

        get hasSagas() {
            return !! this.sagas.length;
        }

        get hasInjectedSagas() {
            return this.state.injectedSagas;
        }

        set hasInjectedSagas( injectedSagas ) {
            this.setState( {injectedSagas} );
        }

        injectSagas() {
            if ( this.hasSagas && ! this.hasInjectedSagas ) {
                this.store.dispatch( runSagas( this.sagas, this.uid ) );
                this.hasInjectedSagas = true;
            }
        }

        render() {
            return (
                <WrappedComponent {...this.props} />
            );
        }
    }

    return SagaInjector;
}

export default curry( SagaInjectorHOC );
