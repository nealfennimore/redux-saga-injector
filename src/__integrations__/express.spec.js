import express from 'express';
import request from 'request';
import createStore, {ShowCount, incrementSaga, delayedIncrementSaga} from '../__fixtures__/store';
import renderer from '../__fixtures__/renderer';
import SagaInjector from '../components/SagaInjector';

const ResponseFactory = ( response )=>{
    let send = resolve => ( ...args )=> {
        resolve( response.send( ...args ) );
    };
    const promise = new Promise( ( resolve )=>{
        send = send( resolve );
    } );
    return {
        promise,
        send,
    };
};

describe( 'Express SSR', ()=>{
    let AppComponent, app, store, server, req, response, res;
    beforeEach( ()=>{
        AppComponent = SagaInjector( {
            sagas: [
                incrementSaga,
                incrementSaga,
                incrementSaga
            ]
        } )( ShowCount );
        req = jest.fn();
        response = {
            send: jest.fn()
        };
        res = ResponseFactory( response );
        store = createStore();
        app = express();
    } );
    afterEach( ()=>{
        server.close();
    } );

    test( 'should increment correctly', async()=>{
        app.get( '*', renderer( AppComponent, store, req, res ) );
        server = app.listen( 3001, '127.0.0.1' );
        request( 'http://localhost:3001' );

        await res.promise;

        expect( response.send.mock.calls ).toMatchSnapshot();
    } );
    test( 'should increment correctly 9 times', async()=>{
        AppComponent = SagaInjector( {
            sagas: [
                incrementSaga,
                incrementSaga,
                incrementSaga,
                incrementSaga,
                incrementSaga,
                incrementSaga,
                incrementSaga,
                incrementSaga,
                incrementSaga,
            ]
        } )( ShowCount );
        app.get( '*', renderer( AppComponent, store, req, res ) );
        server = app.listen( 3001, '127.0.0.1' );
        request( 'http://localhost:3001' );

        await res.promise;

        expect( response.send.mock.calls ).toMatchSnapshot();
    } );
    test( 'should increment correctly when delayed', async()=>{
        AppComponent = SagaInjector( {
            sagas: [
                delayedIncrementSaga,
                delayedIncrementSaga,
                delayedIncrementSaga,
            ]
        } )( ShowCount );
        app.get( '*', renderer( AppComponent, store, req, res ) );
        server = app.listen( 3001, '127.0.0.1' );
        request( 'http://localhost:3001' );

        await res.promise;

        expect( response.send.mock.calls ).toMatchSnapshot();
    } );
} );
