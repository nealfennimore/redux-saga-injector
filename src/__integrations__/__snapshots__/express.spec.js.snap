// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Express SSR components that invoke actions during lifecycle should increment in time 1`] = `
Array [
  Array [
    "
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id=\\"app\\"><div>1</div></div>
                </body>
            </html>
        ",
  ],
]
`;

exports[`Express SSR should increment correctly 1`] = `
Array [
  Array [
    "
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id=\\"app\\"><div>6</div></div>
                </body>
            </html>
        ",
  ],
]
`;

exports[`Express SSR should increment correctly 9 times 1`] = `
Array [
  Array [
    "
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id=\\"app\\"><div>18</div></div>
                </body>
            </html>
        ",
  ],
]
`;

exports[`Express SSR should increment correctly when delayed 1`] = `
Array [
  Array [
    "
            <!DOCTYPE html>
            <html>
                <head>
                </head>
                <body>
                    <div id=\\"app\\"><div>0</div></div>
                </body>
            </html>
        ",
  ],
]
`;
